codegen/src/api/constants.rs

```diff
@@ -48,9 +48,9 @@ pub fn generate_constants(
 
         quote! {
             #( #[doc = #docs ] )*
-            pub async fn #fn_name(&self) -> ::core::result::Result<#return_ty, ::subxt::BasicError> {
+            pub fn #fn_name(&self) -> ::core::result::Result<#return_ty, ::subxt::BasicError> {
                 let locked_metadata = self.client.metadata();
-                let metadata = locked_metadata.lock().await;
+                let metadata = locked_metadata.read();
                 if metadata.constant_hash(#pallet_name, #constant_name)? == [#(#constant_hash,)*] {
```

Comment:
I like the pattern you did in `calls.rs`; I reckon it'd be nice to repeat that here and in `validate_metdata` below for consistency :)

Reply:
Have updated for `validate_metdata` thanks! For `constants.rs` the metadata needs to live a bit longer to extract the constant value from runtime.